### PR TITLE
Increase Paillier key size to 2048

### LINKED ISSUE #55: Update Paillier moduli to be 2048-bits
Currently the Paillier moduli are 1024-bits, which provides insufficient security (80 bits) against known attacks. To fix this issue, update the key generation method for Paillier to produce a 2048-bit modulus instead. This gives us a default security level (for the uses of Paillier in this protocol) of about 112 bits.

### PR DESCRIPTION
**Pull request by marsella**
Resolves #55 

Primary changes:
- Updates the script for pre-generating primes for testing to use OpenSSL's primegen and adds a set of 1000+ 1024-bit primes.
- Updates the code for reading pre-generated primes to cleanly handle errors (e.g. if the pre-generated file doesn't exist).
- Adds a lot of documentation to the parameters module

Review requests
- @indomitableSwan: Please take a look at the parameters documentation.

Follow-up thoughts
With the current `BigNumber` backend, prime generation becomes concerningly slow -- O(2 minutes) to generate a single prime. I tried to run a benchmark on this PR, and it's estimating taking 15 hours to complete a step that took 40 minutes with 512-bit primes. This seems bonkers, and might have been a relative fluke (e.g. the prime generation used to make the estimate was randomly particularly slow), but still concerning.

When I was working on the prime pre-generation script, I found that using OpenSSL to generate primes was 60x faster than `BigNumber`, on average. Running the same benchmark mentioned above for 1024-bit primes is estimated to take 15 minutes using OpenSSL for all big-number operations (including prime gen and modular arithmetic). We should look more closely into the pros and cons of switching to OpenSSL in general.

Even if we don't switch in general, OpenSSL is fast enough at prime generation that we could take out all the reading-primes-from-file optimizations and just do a conditional in the `get_random_safe_prime` function that uses OpenSSL in tests and whatever the library is configured for otherwise.

### CODE REVIEW

File: src/keygen/participant.rs [hunk 1]
```diff
@@ -76,7 +76,8 @@ impl KeygenParticipant {
 
     /// Processes the incoming message given the storage from the protocol
     /// participant (containing auxinfo and keygen artifacts). Optionally
-    /// produces a [KeysharePrivate] and [KeysharePublic] once keygen is
+    /// produces a [`KeysharePrivate`](super::keyshare::KeySharePrivate) and
+    /// [`KeySharePublic`](super::keyshare::KeySharePublic) once keygen is
```
Comment (marsella) [inline src/keygen/participant.rs:80]: note: I build the doc locally with `--document-private-items` and had to fix this to make it work. Should we change the CI to build private docs by default? Most of the crate is private but I think it's good practice in a cryptography library to tend toward good, voluminous documentation.

File: src/paillier.rs [hunk 2]
```diff
@@ -131,50 +131,62 @@ pub(crate) mod prime_gen {
...
+                .filter(|prime| prime.bit_length() == PRIME_BITS)
+                .collect(),
+
+            // This is the recommended `PRIME_BITS` value.
+            1024 => crate::safe_primes_1024::SAFE_PRIMES
+                .iter()
+                .map(|s| BigNumber::from_slice(hex::decode(s).unwrap()))
+                .collect(),
+
+            _ => Vec::new(),
```
Comment (tyurek) [inline src/paillier.rs:163]: Is it worth replacing this with a more explicit error? Otherwise, changing PRIME_BITS causes tests to fail in a more opaque way (thread 'my_test_name' panicked at 'cannot sample empty range')

File: src/parameters.rs [hunk 3]
```diff
@@ -5,17 +5,61 @@
...
+///
+/// The product of two 1024-bit primes will produce 2048-bit Pallier moduli.
+pub(crate) const PRIME_BITS: usize = 1024;
+
+/// Number of repetitions required for statistical security in proofs that allow an adversary to
+/// guess a challenge value correctly with probability 1/2.
+///
+/// Historical documentation note without a verified source:
+/// > Needs to be a multiple of 8 for Pi_prm
+pub(crate) const SOUNDNESS_PARAMETER: usize = 8;
```
Comment (tyurek) [inline src/parameters.rs:65]: I haven't yet figured out where the multiple of 8 part comes from, but it looks like SOUNDNESS_PARAMETER should be 80 instead of 8? (can be addressed later)

**Review by tyurek — CHANGES_REQUESTED** [review]
Minor comments only. This looks really good, I especially like the level of detail added to parameters.rs!

File: examples/gen_primes.rs [hunk 4]
```diff
@@ -7,42 +7,118 @@
...
+
+/// Use the BigNumber crate to generate a safe prime and encode as a hex string.
+#[allow(unused)]
+fn bignumber_safe_prime() -> String {
+    let prime = BigNumber::safe_prime(PRIME_SIZE);
+    hex::encode(prime.to_bytes())
+}
+
+/// Use the openssl crate to generate a safe prime and encode.
+/// This is way faster than BigNumber.
```
Comment (indomitableSwan) [inline examples/gen_primes.rs:95]: Do you mean Rust's BigInt library here?

File: src/keygen/participant.rs (hunk 1 above)
Comment (indomitableSwan) [reply src/keygen/participant.rs:80]: That seems like a reasonable approach to me.

File: src/parameters.rs [hunk 5]
```diff
@@ -5,17 +5,61 @@
...
-///////////////
-// Constants //
-// ========= //
-///////////////
+//! Security parameter definitions.
+//!
+//! This module defines security parameters that ensure the protocol satisfies
+//! 80-bit statistical security and 112-bit computational security (the paper claims
+//! 128-bit computational security, but NIST's Recommendation for Key Management \[1\]
+//! suggests this might be an overestimation due to the Paillier modulus choice).
```
Comment (indomitableSwan) [inline src/parameters.rs:13]: Why do we sound unsure of this?

File: src/parameters.rs (hunk 3 above)
Comment (indomitableSwan) [reply src/parameters.rs:65]: Yes, 8 must have been a temporary choice for convenience or something.

**Review by indomitableSwan — APPROVED** [review]
LGTM. Just a couple of minor comments about documentation that you may want to fix.

And we should fix the soundness parameter in its own ticket, right?

**Comment by indomitableSwan** [comment]
> Resolves #55
> 
> Primary changes:
> 
> * Updates the script for pre-generating primes for testing to use OpenSSL's primegen and adds a set of 1000+ 1024-bit primes.
> * Updates the code for reading pre-generated primes to cleanly handle errors (e.g. if the pre-generated file doesn't exist).
> * Adds a lot of documentation to the parameters module
> 
> Review requests
> 
> * @indomitableSwan: Please take a look at the parameters documentation.
> 
> Follow-up thoughts With the current `BigInt` backend, prime generation becomes concerningly slow -- O(2 minutes) to generate a single prime. I tried to run a benchmark on this PR, and it's estimating taking 15 hours to complete a step that took 40 minutes with 512-bit primes. This seems bonkers, and might have been a relative fluke (e.g. the prime generation used to make the estimate was randomly particularly slow), but still concerning.
> 
> When I was working on the prime pre-generation script, I found that using OpenSSL to generate primes was 60x faster than `BigInt`, on average. Running the same benchmark mentioned above for 1024-bit primes is estimated to take 15 minutes using OpenSSL for all big-number operations (including prime gen and modular arithmetic). We should look more closely into the pros and cons of switching to OpenSSL in general.
> 
> Even if we don't switch in general, OpenSSL is fast enough at prime generation that we could take out all the reading-primes-from-file optimizations and just do a conditional in the `get_random_safe_prime` function that uses OpenSSL in tests and whatever the library is configured for otherwise. However, installing OpenSSL on the CI every time might cancel out any speedup.

I don't think it's unreasonable to rely on a wrapper around OpenSSL, and it makes sense to me that OpenSSL would have spent time implementing optimizations for safe prime generation. It is a little surprising that Rust's BigInt library is so slow at safe prime generation by comparison, though.

File: examples/gen_primes.rs [hunk 6]
```diff
@@ -7,42 +7,118 @@
...
+use openssl::{self, bn::BigNum, error::ErrorStack};
 use std::{
     fs::OpenOptions,
     io::prelude::*,
     time::{Duration, Instant},
 };
 
-/// Used to pre-generate 512-bit safe primes, so that they can
-/// be stored in a text file to make testing this library faster
+const PRIME_SIZE: usize = 1024;
```
[the remark this answers is not in the document]
Comment (marsella) [reply examples/gen_primes.rs:17]: Ah, I did the prime generation before I made the change in the crate, so didn't try to re-use. It does seem more misuse-resistant to have a single value, but it would require making a bunch of things more public (e.g. the parameters module) and I'm not sure I want to mess with that without thinking carefully about our intended API.

File: src/parameters.rs (hunk 3 above)
Comment (marsella) [reply src/parameters.rs:65]: Yeah, I agree. Made #73 to fix.

File: examples/gen_primes.rs (hunk 4 above)
Comment (marsella) [reply examples/gen_primes.rs:95]: Good catch, clarified.

File: src/parameters.rs (hunk 5 above)
Comment (marsella) [reply src/parameters.rs:13]: Well, there are other recommendations that say 2048 still gives you 128 bits of security, but I will rephrase with confidence.

File: src/paillier.rs (hunk 2 above)
Comment (marsella) [reply src/paillier.rs:166]: I made some changes and now tests that call `get_prime_pair_from_pool_insecure` directly get a more meaningful error, but those that call `PaillierDecryptionKey::new` fail with a `RetryFailed` error. I'm going to say that gives enough info for now, but we might revisit in the future.

Good catch, though. I didn't realize the `rng` function panics with a bad range.

File: src/keygen/participant.rs (hunk 1 above)
Comment (marsella) [reply src/keygen/participant.rs:80]: Added #79

File: src/parameters.rs (hunk 5 above)
Comment (indomitableSwan) [reply src/parameters.rs]: Oh? Which ones are you thinking of?

Comment (marsella) [reply src/parameters.rs]: Hm, I might have misinterpreted the table at keylength.com. I think I parsed the "symmetric" column as the overall computational security level, which would imply that ANSSI says a 2048 bit factoring modulus gives 128 bits of security. I took a look at that reference and, although it's in French, I couldn't see a specific computational security level that 2048 bits provides you -- it just says that the minimum size allowed should be 2048.

File: examples/gen_primes.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (tyurek) [reply examples/gen_primes.rs]: Ah okay. Yeah, we probably don't need to export that then

**Review by tyurek — APPROVED** [review]
LGTM
